umbraco/Umbraco.CMS.Backoffice
# Feature: Persisting User Permissions

### Persisting User Permissions:

Note: See server implementation details here: https://github.com/umbraco/Umbraco-CMS/pull/15734

The PR wraps up the final parts of persisting the entity and granular permissions on a user group.

**Currently, we allow two extension points on the client for user permissions:**
- **Entity User Permissions** - Relates to an entity (ex document). 
- **Granular User Permission** - Relates to a $type server schemaType.

Each permission comes with a set of verbs, that will be checked against client and server side (client side check will be implemented in the next PR).

The Core currently ships with entity user permission for documents. The permissions are as follows:

|Current Backoffice Letter |  Verb|
| ------------- | ------------- |
| C| Umb.Document.Create| 
| F| Umb.Document.Read| 
| A| Umb.Document.Update| 
| D| Umb.Document.Delete| 
| I| Umb.Document.CreateBlueprint| 
| N| Umb.Document.Notifications| 
| U| Umb.Document.Publish| 
| R| Umb.Document.Permissions| 
| Z| Umb.Document.Unpublish| 
| O| Umb.Document.Duplicate| 
| M| Umb.Document.Move| 
| S| Umb.Document.Sort| 
| I| Umb.Document.CultureAndHostnames| 
| P| Umb.Document.PublicAccess| 
| K| Umb.Document.Rollback| 
| V| Umb.DocumentRecycleBin.Restore| 

Entity User Permissions will be registered in the extension registry with a manifest with the following type. Ex:
```json
{
    "type": "entityUserPermission",
    "alias": "Umb.UserPermission.Document.Rollback",
    "name": "Document Rollback User Permission",
    "meta": {
      "entityType": "document",
      "verbs": ["Umb.Document.Rollback"],
      "labelKey": "actions_rollback",
      "descriptionKey": "actionDescriptions_rollback",
      "group": "administration",
    },
  },
```

Granular permissions will also be registered. It is possible to provide a custom element to build the needed UX for that type of permission:
```json
{
    "type": "userGranularPermission",
    "alias": "Umb.UserGranularPermission.Document",
    "name": "Document Granular User Permission",
    "element": "element.js",
    "meta": {
      "schemaType": "DocumentPermissionPresentationModel",
      "label": "Documents",
      "description": "Assign permissions to specific documents",
    },
  },
```

Entity User Permissions in the UI:
<img width="934" alt="Screenshot 2024-02-26 at 14 41 13" src="https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/6078361/f74de06d-d2af-4d5b-9af0-b3a4d573c988">

Granular User Permissions in the UI:
<img width="892" alt="Screenshot 2024-02-26 at 14 42 43" src="https://github.com/umbraco/Umbraco.CMS.Backoffice/assets/6078361/618ee75e-9df9-47ed-aaba-6273218a33a7">

**What to test:**
* Make sure you can set default permissions on a user group
* Make sure you can add, update, and remove granular permissions on a user group
